README.md
Fast 3x3 SVD
===========

This is an implementation of the method described in <a href="http://pages.cs.wisc.edu/~sifakis/papers/SVD_TR1690.pdf">"Computing the Singular Value Decomposition of 3x3 matrices with minimal branching and elementary floating point operations"</a>. I implemented this as part of <a href="http://wyegelwel.github.io/snow/">a group project</a> for a computer graphics course. 

Execution time per svd call on the CPU is about 2.0 microseconds. Tested on a AMD Phenom(tm) II X4 965 Processor. 

Execution time on the GPU is about 174 microseconds. Tested on a NVIDIA GeForce GTX 460 (profiled using nvvp).

Also included are routines for diagonalization / QR decomposition of 3x3 matrices, which may be useful in their own right. 


## Usage

Just include the header file and you are good to go! 

```C++

#include "svd3/svd3.h"
float a11, a12, a13, a21, a22, a23, a31, a32, a33;

a11= -0.558253; a12 = -0.0461681; a13 = -0.505735;
a21 = -0.411397; a22 = 0.0365854; a23 = 0.199707;
a31 = 0.285389; a32 =-0.313789; a33 = 0.200189;

float 	u11, u12, u13, 
		u21, u22, u23, 
		u31, u32, u33;

float 	s11, s12, s13, 
		s21, s22, s23, 
		s31, s32, s33;

float 	v11, v12, v13, 
		v21, v22, v23, 
		v31, v32, v33;

svd(a11, a12, a13, a21, a22, a23, a31, a32, a33,
    u11, u12, u13, u21, u22, u23, u31, u32, u33,
    s11, s12, s13, s21, s22, s23, s31, s32, s33,
    v11, v12, v13, v21, v22, v23, v31, v32, v33);

```

See the included Mathematica notebook for derivations of numerical shortcuts.

## License
MIT License, Eric V. Jang 2014

## Building

See [INSTALL](INSTALL.md) text file.

Continuous integration:
 - [![Build Status](https://travis-ci.org/EncovGroup/svd3.svg?branch=develop)](https://travis-ci.org/EncovGroup/svd3) develop branch.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.0.0] - 2018-08-23
### Added
- CHANGELOG.md to keep track of changes wrt the original version
- INSTALL.md to explain building instructions
- LICENSE.md
- CMake files for building

### Changed
- moved all files in a more standard project tree
- renamed `main.cpp` into `main_gpu.cpp`

### Removed
- old makefiles

src/samples/main_cpu.cpp
#include <stdio.h>
#include <svd3/svd3.h>
#include <time.h>

// some printing utilities
inline void printMat3(float a11, float a12, float a13,
               float a21, float a22, float a23,
               float a31, float a32, float a33)
{
    printf("%f %f %f \n", a11, a12, a13);
    printf("%f %f %f \n", a21, a22, a23);
    printf("%f %f %f \n", a31, a32, a33);
}

inline void printQuat(float * q)
{
    // print w,x,y,z
    printf("%f %f %f %f\n",q[3],q[0],q[1],q[2]);
}


int main(void)
{
    float a11, a12, a13, a21, a22, a23, a31, a32, a33;

    a11= -0.558253; a12 = -0.0461681; a13 = -0.505735;
    a21 = -0.411397; a22 = 0.0365854; a23 = 0.199707;
    a31 = 0.285389; a32 =-0.313789; a33 = 0.200189;

    // printf("Original Matrix:\n");
    // printMat3(a11, a12, a13, a21, a22, a23, a31, a32, a33);

    float u11, u12, u13, u21, u22, u23, u31, u32, u33;
    float s11, s12, s13, s21, s22, s23, s31, s32, s33;
    float v11, v12, v13, v21, v22, v23, v31, v32, v33;

    clock_t start, end;
    start = clock();
    for (int i=0; i<1e6; i++)
    {
        svd(a11, a12, a13, a21, a22, a23, a31, a32, a33,
        u11, u12, u13, u21, u22, u23, u31, u32, u33,
        s11, s12, s13, s21, s22, s23, s31, s32, s33,
        v11, v12, v13, v21, v22, v23, v31, v32, v33);
    }    
    end = clock();

    printf("Average SVD takes %f microseconds \n ", 1e6*(double(end - start) / 1e6 / CLOCKS_PER_SEC ) );
    
    printf("U:\n");
    printMat3(u11, u12, u13, u21, u22, u23, u31, u32, u33);
    printf("S:\n");
    printMat3(s11, s12, s13, s21, s22, s23, s31, s32, s33);
    printf("V:\n");
    printMat3(v11, v12, v13, v21, v22, v23, v31, v32, v33);

    float t11, t12, t13, t21, t22, t23, t31, t32, t33;
    multAB(u11, u12, u13, u21, u22, u23, u31, u32, u33,
           s11, s12, s13, s21, s22, s23, s31, s32, s33,
           t11, t12, t13, t21, t22, t23, t31, t32, t33);

    float m11, m12, m13, m21, m22, m23, m31, m32, m33;
    multAB(t11, t12, t13, t21, t22, t23, t31, t32, t33,
           v11, v21, v31, v12, v22, v32, v13, v23, v33,
           m11, m12, m13, m21, m22, m23, m31, m32, m33);

    printf("USV* : \n");
    printMat3(m11, m12, m13, m21, m22, m23, m31, m32, m33);

    return 0;

}



src/samples/main_glm.cpp
#include <iostream>
#include <glm/geometric.hpp>
#include <glm/gtc/epsilon.hpp>
#include <glm/gtc/quaternion.hpp>
#include <glm/gtx/quaternion.hpp>
#include <glm/gtx/fast_square_root.hpp>
#include <glm/gtx/fast_trigonometry.hpp>
#include <glm/gtc/matrix_access.hpp>
#include <math.h>

#define _sqrtHalf 0.70710678
#define _gamma 5.828427124 // FOUR_GAMMA_SQUARED = sqrt(8)+3;
#define _cstar 0.923879532 // cos(pi/8)
#define _sstar 0.3826834323 // sin(p/8)

#define _EPSILON_ 1e-6
#define EPSILON _EPSILON_

#define EQ(a, b) (fabs((a) - (b)) < _EPSILON_)
#define NEQ(a, b) (fabs((a) - (b)) > _EPSILON_)

void printMat3(glm::mat3 mat) {
    // prints by rows
    for (int j=0; j<3; ++j) // g3d stores column-major
    {
        for (int i=0; i<3; ++i)
        {
            printf("%f   ", mat[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

void printQuat(glm::quat q)
{
    std::cout << q.w << "  " << q.x << "  " << q.y  << "  " << q.z <<std::endl;
    //std::cout << q[3] << "  " << q[0] << "  " << q[1]  << "  " << q[2] <<std::endl;
}

inline float accurateRSQRT(float x)
{
    // used in step 3
    // cool result from Lomont 2003
    float y = glm::fastSqrt(x);
    return y * (3-x*y*y)/2;
}

inline float accurateSQRT(float x) { return x * accurateRSQRT(x); } // used in step 3

void condSwap(bool c, float &X, float &Y)
{
    // used in step 2
    float Z = X;
    X = c ? Y : X;
    Y = c ? Z : Y;
}

// swapping functions for entire rows
void condSwap(bool c, glm::vec3 &X, glm::vec3 &Y)
{
    // used in step 2
    glm::vec3 Z = X;
    X = c ? Y : X;
    Y = c ? Z : Y;
}

void condNegSwap(bool c, glm::vec3 &X, glm::vec3 &Y)
{
    // used in step 2 and 3
    glm::vec3 Z = -X;
    X = c ? Y : X;
    Y = c ? Z : Y;
}

void condNegSwap(bool c, int c1, int c2, glm::quat &qV)
{
    // condNegSwap can be modified to operate well
    // on quaternion representation of V
    glm::quat qR;
    // qR = (1, 0, 0, c) for 1,2
    // what are the other quaternions corresponding to the other rot matrices?
    qV = qV * qR;
}

//void approximateGivensAngles(float a11, float a12, float a22, float &c, float &s)
//{
//    // this function is unused
//    // problem, second round the angles are flipped?
//    /*
//     * approximates givens angles for 2x2 submatrix. used in step 1
//     *
//     * A = (a11 a12
//     *      a12 a22)
//     *
//     * where 1,2 correspond to p,q values for givens rotation of a 3x3 matrix
//     */
//    float A2 = a12 * a12;
//    float B2 = (a11-a22)*(a11-a22);
//    bool b = A2 < B2;
//    //float w = rsqrt(A2+B2);
//    float w = glm::fastInverseSqrt(A2+B2);
//    s = b ? w*a12 : _sqrtHalf;
//    c = b ? w*(a11-a22) : _sqrtHalf;
//    // we can represent this rotation with quaternion
//    // (cos(theta/2),0,0,sin(theta/2)) for p,q = 1,2
//    // up to the user to build the appropriate quaternion or rotation matrix
//}

void approximateGivensQuaternion(float a11, float a12, float a22, float &ch, float &sh)
{
    /*
     * Given givens angle computed by approximateGivensAngles,
     * compute the corresponding rotation quaternion.
     */
    // used during second half of each jacobi iteration step?
    ch = 2*(a11-a22);
    sh = a12;
    bool b = _gamma*sh*sh < ch*ch;
    float w = glm::fastInverseSqrt(ch*ch+sh*sh);
    ch=b?w*ch:_cstar;
    sh=b?w*sh:_sstar;
    //qV=glm::quat(ch,0,0,sh)
    //up to the user to build the appropriate quaternion or unscaled rotation matrix
}

void jacobiConjugation(int p, int q, glm::mat3 &S, glm::quat &qV)
{
    //std::cout << "matrix before conjugation" << std::endl;
    //printMat3(S);
    // eliminate off-diagonal entries Spq, Sqp
    float ch,sh;
    approximateGivensQuaternion(S[0][0],S[1][0],S[1][1],ch,sh);

    // build rotation matrix Q
    glm::mat3 Q;
    float scale = ch*ch+sh*sh;
    float a = (ch*ch-sh*sh)/scale;
    float b = (2*sh*ch)/scale;
    
//    printf("a=%f, b=%f \n", a, b );

    Q[0][0] = a;    Q[1][0] = -b;
    Q[0][1] = b;    Q[1][1] = a;

    // perform the conjugation to annihilate S = Q' S Q
    S =glm::transpose(Q) * S * Q;


    // std::cout << "result of conjugation" << std::endl;
    // printMat3(S);
    // update cumulative rotation quaternion qV
    // tmp1, tmp2, tmp3
    glm::vec3 tmp(qV.x,qV.y,qV.z);
    tmp *= sh;
    sh *= qV.w;
    // printf("sh=%f\n",sh);
    // original
    qV *= ch;

    // this arranges such that for (p,q) = ((0,1),(1,2),(0,2)),
    // n = (0,1,2)
    int n = 2*q-p-2; // worst hack in the world

    int x = n;
    int y = (n+1)%3;
    int z = (n+2)%3;

    qV[z] += sh;
    qV.w -= tmp[z];
    qV[x] += tmp[y];
    qV[y] -= tmp[x];

    // re-arrange matrix for next iteration
    S = glm::mat3(S[1][1], S[1][2], S[1][0],
                  S[1][2], S[2][2], S[0][2],
                  S[1][0], S[0][2], S[0][0]);
}

void sortSingularValues(glm::mat3 &B, glm::mat3 &V)
{
    // used in step 2
    glm::vec3 b1 = glm::column(B,0); glm::vec3 v1 = glm::column(V,0);
    glm::vec3 b2 = glm::column(B,1); glm::vec3 v2 = glm::column(V,1);
    glm::vec3 b3 = glm::column(B,2); glm::vec3 v3 = glm::column(V,2);
    float rho1 = glm::length2(b1);
    float rho2 = glm::length2(b2);
    float rho3 = glm::length2(b3);
    bool c;

    c = rho1 < rho2;
    condNegSwap(c,b1,b2); condNegSwap(c,v1,v2);
    condSwap(c,rho1,rho2);

    c = rho1 < rho3;
    condNegSwap(c,b1,b3); condNegSwap(c,v1,v3);
    condSwap(c,rho1,rho3);

    c = rho2 < rho3;
    condNegSwap(c,b2,b3); condNegSwap(c,v2,v3);

    // re-build B,V
    B = glm::mat3(b1,b2,b3);
    V = glm::mat3(v1,v2,v3);
}


void QRGivensQuaternion(float a1, float a2, float &ch, float &sh)
{
    // a1 = pivot point on diagonal
    // a2 = lower triangular entry we want to annihilate
    float epsilon = EPSILON;

    float rho = sqrt(a1*a1 + a2*a2);

    sh = rho > epsilon ? a2 : 0;
    ch = fabs(a1) + fmax(rho,epsilon);
    bool b = a1 < 0;
    condSwap(b,sh,ch);
    float w = glm::inversesqrt(ch*ch+sh*sh);
    //float w = glm::fastInverseSqrt(ch*ch+sh*sh);
    ch *= w;
    sh *= w;
}

void QRDecomposition(glm::mat3 B, glm::mat3 &Q, glm::mat3 &R)
{
    // QR decomposition of 3x3 matrices using Givens rotations to
    // eliminate elements B21, B31, B32

    glm::quat qQ; // cumulative rotation
    glm::quat qU; // each Givens rotation in quaternion form
    glm::mat3 U;
    float ch, sh;
    QRGivensQuaternion(B[0][0],B[0][1],ch,sh);
    qU = glm::quat(ch,0,0,sh);
    U = glm::toMat3(qU);
    B = glm::transpose(U) * B;

    // update cumulative rotation
    qQ *= qU;

    // second givens rotation
    QRGivensQuaternion(B[0][0],B[0][2],ch,sh);
    qU = glm::quat(ch,0,-sh,0);
    U = glm::toMat3(qU);
    B = glm::transpose(U) * B;
    qQ *= qU;

    // third Givens rotation
    QRGivensQuaternion(B[1][1],B[1][2],ch,sh);
    qU = glm::quat(ch,sh,0,0);
    U = glm::toMat3(qU);
    B = glm::transpose(U) * B;
    qQ *= qU;
    // B has been transformed into R
    R = B;
    // qQ now contains final rotation for Q
    Q = glm::toMat3(qQ);
}

void jacobiEigenanalysis(glm::mat3 &S, glm::quat &qV)
{
    // wrapper function for the first step
    // solve symmetric eigenproblem using jacobi iteration
    // given a symmetric matrix S, diagonalize it
    // also returns the cumulative rotation as a quaternion
    qV = glm::quat(1,0,0,0);
    float scale;
    int p; int q;
    for(int sweep=0;sweep<4;sweep++)
    {
        // we wish to eliminate the maximum off-diagonal element
        // on every iteration, but cycling over all 3 possible rotations
        // in fixed order (p,q) = (1,2) , (2,3), (1,3) still has
        //  asymptotic convergence
        jacobiConjugation(0,1,S,qV);
        jacobiConjugation(1,2,S,qV);
        jacobiConjugation(0,2,S,qV);
    }
}


int main()
{

    // run a simpe test
    glm::mat3 A;
    // GLM stores matrices in column-major order so this initialization is the transpose of what we want
    A = glm::mat3(   -0.558253,    -0.0461681,    -0.505735,
                   -0.411397  ,   0.0365854  ,   0.199707,
                    0.285389  ,   -0.313789  ,   0.200189);

    A = glm::transpose(A);

    std::cout << "original matrix" << std::endl;
    printMat3(A);
/// 2. Symmetric Eigenanlysis
    // normal equations matrix
    glm::mat3 S = glm::transpose(A) * A;
    // std::cout << "normal equations matrix" << std::endl;
    // printMat3(S);
    glm::quat qV;
    jacobiEigenanalysis(S,qV);
    std::cout << "final S (diagonalized) " << std::endl;
    printMat3(S);
    std::cout << "cumulative rotation quaternion" << std::endl;
    printQuat(qV);


    glm::mat3 V = glm::toMat3(qV); // normalize qV, convert it to matrix
    //std::cout << "final rot matrix V" << std::endl;
    //printMat3(V);
    glm::mat3 B = A * V; // right-multiply A with V => left multiply for column major
    std::cout << "B=AV" << std::endl;
    //printMat3(B);
///// 3. Sorting the singular values (find V)
    sortSingularValues(B,V);
    //std::cout << "sorted B=AV" << std::endl;
    //printMat3(B);
    std::cout << "sorted V" << std::endl;
    printMat3(V);

//    // if columns 2-3 swapped, also update quaternion representation (i dont think we need to though since we're not tracking quats)
///// 4. QR factorization using Givens rotations (find U,S from B=AV)
    glm::mat3 U;
    glm::mat3 Sigma;
    QRDecomposition(B,U,Sigma);
    std::cout << "U" << std::endl;
    printMat3(U);
    std::cout << "Sigma" << std::endl;
    printMat3(Sigma);
    glm::mat3 USV = U * Sigma * glm::transpose(V);
    std::cout << "product USV'"<< std::endl;
    printMat3(USV);

    return 0;
}

